File: server/audit/scoring.py
"""Scoring utilities — normalize metrics to 0-100, pick traffic light,
compute maintainability 60/40 mix, aggregate global score.
"""

from __future__ import annotations

from typing import Iterable

from .schema import (
    CharacteristicResult,
    Finding,
    Severity,
    SquareCharacteristic,
    TrafficLight,
)


# Traffic light thresholds (score out of 100)
GREEN_MIN = 80.0
AMBER_MIN = 60.0


def traffic_light(score: float) -> TrafficLight:
    if score >= GREEN_MIN:
        return TrafficLight.GREEN
    if score >= AMBER_MIN:
        return TrafficLight.AMBER
    return TrafficLight.RED


def clamp(value: float, lo: float = 0.0, hi: float = 100.0) -> float:
    return max(lo, min(hi, value))


# Severity → points deducted from a starting score of 100
SEVERITY_PENALTY: dict[Severity, float] = {
    Severity.CRITICAL: 25.0,
    Severity.HIGH: 12.0,
    Severity.MEDIUM: 5.0,
    Severity.LOW: 2.0,
    Severity.INFO: 0.5,
}


def score_from_findings(findings: Iterable[Finding], base: float = 100.0) -> float:
    """Deduct points from `base` per finding severity. Floor at 0."""
    score = base
    for f in findings:
        score -= SEVERITY_PENALTY.get(f.severity, 1.0)
    return clamp(score)


def ratio_to_score(ratio: float, invert: bool = False) -> float:
    """Normalize a 0..1 ratio to 0..100. If `invert`, higher ratio is worse."""
    r = max(0.0, min(1.0, ratio))
    return clamp((1 - r) * 100 if invert else r * 100)


# ------ Maintainability 60/40 mix ------

CLASSIC_WEIGHT = 0.60
SPECBOX_WEIGHT = 0.40


def maintainability_score(classic_score: float, specbox_score: float) -> tuple[float, dict]:
    """Combine classic industry metrics (60%) with SpecBox MCP signals (40%).

    Returns (final_score, breakdown_dict). The breakdown MUST appear in the
    report so readers can see exactly how the score was computed.
    """
    final = (classic_score * CLASSIC_WEIGHT) + (specbox_score * SPECBOX_WEIGHT)
    breakdown = {
        "classic_60": {
            "score": round(classic_score, 2),
            "weight": CLASSIC_WEIGHT,
            "contribution": round(classic_score * CLASSIC_WEIGHT, 2),
        },
        "specbox_40": {
            "score": round(specbox_score, 2),
            "weight": SPECBOX_WEIGHT,
            "contribution": round(specbox_score * SPECBOX_WEIGHT, 2),
        },
        "formula": "0.60 * classic + 0.40 * specbox",
    }
    return clamp(final), breakdown


# ------ Global aggregation ------

def global_score(results: list[CharacteristicResult]) -> float:
    """Simple mean of non-skipped characteristic scores.

    Skipped characteristics do not penalize the project (reported in meta).
    """
    active = [r.score for r in results if not r.skipped]
    if not active:
        return 0.0
    return clamp(sum(active) / len(active))


def ordered_characteristics(
    results_by_id: dict[SquareCharacteristic, CharacteristicResult],
) -> list[CharacteristicResult]:
    """Return characteristics in the canonical SQuaRE order for reports."""
    from .schema import SQUARE_ORDER

    out: list[CharacteristicResult] = []
    for c in SQUARE_ORDER:
        if c in results_by_id:
            out.append(results_by_id[c])
    return out

File: .quality/scripts/audit/lib/scoring.mjs
/**
 * lib/scoring.mjs — Client-side mirror of server/audit/scoring.py (UC-663).
 *
 * Pure functions, no I/O. Kept numerically identical to the Python so a report
 * built on the client scores the same as the legacy server-side audit did.
 */

import { Severity, TrafficLight } from './schema.mjs';

// Traffic light thresholds (score out of 100) — mirror scoring.py.
export const GREEN_MIN = 80.0;
export const AMBER_MIN = 60.0;

export function trafficLight(score) {
  if (score >= GREEN_MIN) return TrafficLight.GREEN;
  if (score >= AMBER_MIN) return TrafficLight.AMBER;
  return TrafficLight.RED;
}

export function clamp(value, lo = 0.0, hi = 100.0) {
  return Math.max(lo, Math.min(hi, value));
}

// Severity → points deducted from a starting score (mirror SEVERITY_PENALTY).
export const SEVERITY_PENALTY = Object.freeze({
  [Severity.CRITICAL]: 25.0,
  [Severity.HIGH]: 12.0,
  [Severity.MEDIUM]: 5.0,
  [Severity.LOW]: 2.0,
  [Severity.INFO]: 0.5,
});

/**
 * Deduct points from `base` per finding severity. Floor at 0.
 * @param {Array<{severity:string}>} findings
 * @param {number} [base]
 */
export function scoreFromFindings(findings, base = 100.0) {
  let score = base;
  for (const f of findings) {
    score -= SEVERITY_PENALTY[f.severity] ?? 1.0;
  }
  return clamp(score);
}

/** Normalize a 0..1 ratio to 0..100. If invert, higher ratio is worse. */
export function ratioToScore(ratio, invert = false) {
  const r = Math.max(0.0, Math.min(1.0, ratio));
  return clamp(invert ? (1 - r) * 100 : r * 100);
}

// ------ Maintainability 60/40 mix ------

export const CLASSIC_WEIGHT = 0.6;
export const SPECBOX_WEIGHT = 0.4;

/**
 * Combine classic industry metrics (60%) with SpecBox MCP signals (40%).
 * @returns {[number, object]} [finalScore, breakdown]
 */
export function maintainabilityScore(classicScore, specboxScore) {
  const final = classicScore * CLASSIC_WEIGHT + specboxScore * SPECBOX_WEIGHT;
  const r2 = (x) => Math.round((x + Number.EPSILON) * 100) / 100;
  const breakdown = {
    classic_60: {
      score: r2(classicScore),
      weight: CLASSIC_WEIGHT,
      contribution: r2(classicScore * CLASSIC_WEIGHT),
    },
    specbox_40: {
      score: r2(specboxScore),
      weight: SPECBOX_WEIGHT,
      contribution: r2(specboxScore * SPECBOX_WEIGHT),
    },
    formula: '0.60 * classic + 0.40 * specbox',
  };
  return [clamp(final), breakdown];
}

// ------ Global aggregation ------

/** Simple mean of non-skipped characteristic scores (mirror global_score). */
export function globalScore(results) {
  const active = results.filter((r) => !r.skipped).map((r) => r.score);
  if (active.length === 0) return 0.0;
  return clamp(active.reduce((a, b) => a + b, 0) / active.length);
}
